### ISSUE TITLE
Some questions about training kicking behavior

### ISSUE THREAD

**Issue by zkk2157** [m-abr/FCPCodebase#30]
I am currently writing code to train the behavior of kicking a ball. Is there any code example I can refer to? Also, in the reset() function, I put the ball 0.5 meters in front of the agent. When the agent starts walking, the ball returns to the middle of the court. What is the problem?In the code for training the kicking action, I divided it into two steps, the first is to walk towards the ball, and the second is to perform the kicking action. In the step() function, when walking towards the ball, do I need to return obs, reward, terminal,{}?

**Comment by m-abr** [comment]
### Handling the ball resetting to the middle of the field
This behavior is likely caused by the referee interfering with the ball's position. As explained in the documentation ([link here](https://docs.google.com/document/d/1aJhwK2iJtU-ri_2JOB8iYvxzbPskJ8kbk_4rb3IK3yc/edit?tab=t.0#heading=h.txmrqxm6q5b3)), under Section *10.4.1 Referee interference*, you can handle this issue by choosing one of the following options:

1. **Turning off the referee** through the server’s configuration parameters (though note that this may not currently work due to a server bug).
2. **Setting the play mode to `GameOver`** using `self.player.scom.unofficial_set_play_mode("GameOver")` at the end of the gym’s `__init__` method.
3. **Setting the play mode to `PlayOn`** using `self.player.scom.unofficial_set_play_mode("PlayOn")` at the end of the gym’s `__init__` method, and then frequently resetting the game time (e.g., at every step) using `self.player.scom.unofficial_set_game_time(0)`. This option checks for self-collisions, which can be useful for certain behaviors (e.g. kicking the ball). However, note that the referee will still intervene in cases like the ball going out of bounds or entering a goal.

### Code example to refer to
You can use `/scripts/gyms/Basic_Run.py` as a base behavior for your implementation.

### Dividing training into two steps
Your approach of splitting the training into two steps (walking towards the ball and then kicking) is a good idea. However, it's crucial to organize your `reset()` and `step()` functions correctly:

1. **`reset()` Function**:
   - In the `reset()` function, you should handle all the non-learned behaviors:
     - Place the ball at a specific position.
     - Program the robot to walk towards the ball.
     - End the `reset()` function when the robot is close enough to kick the ball.
   - This ensures that the learning process focuses solely on the kicking part during the `step()` function.

2. **`step()` Function**:
   - Here is a draft implementation of the `step()` function, assuming an action size of 16 (defined in `__init__` as `self.no_of_actions = act_size = 16`):

     ```python
     def step(self, action):
         w = self.player.world
         r = self.player.world.robot

         # Example rescaling of the action outputs for a kick
         r.joints_target_speed[2:18] = action * [6, 6, 3, 3, 6, 6, 6, 6, 6, 6, 2, 2, 1, 1, 1, 1]
         
         self.sync()  # Run simulation step
         self.step_counter += 1
         
         if self.step_counter > 20:  # Allow 0.4 seconds for the kick (adjust as needed)
             # Stop controlling the robot and evaluate the kick
             while True:
                 # (CODE TO EXIT LOOP WHEN THE BALL STOPS)
                 r.behavior.execute("Zero_Bent_Knees")  # Example idle behavior
                 self.sync()  # Run simulation step

             # (CODE TO CALCULATE YOUR_REWARD)            

             # Terminate the episode and return the reward
             return self.observe(), YOUR_REWARD, True, {}
         
         else:  # No reward until the end
             return self.observe(), 0, False, {}
     ```

### Adjusting the number of steps per environment
Since the episodes in your environment are very short, you might want to adjust the `n_steps_per_env` hyperparameter in the `train()` function of the `Train` class. For example, you can set:

```python
n_steps_per_env = 128

**Comment by zkk2157** [comment]
> ### Handling the ball resetting to the middle of the field
> This behavior is likely caused by the referee interfering with the ball's position. As explained in the documentation ([link here](https://docs.google.com/document/d/1aJhwK2iJtU-ri_2JOB8iYvxzbPskJ8kbk_4rb3IK3yc/edit?tab=t.0#heading=h.txmrqxm6q5b3)), under Section _10.4.1 Referee interference_, you can handle this issue by choosing one of the following options:
> 
> 1. **Turning off the referee** through the server’s configuration parameters (though note that this may not currently work due to a server bug).
> 2. **Setting the play mode to `GameOver`** using `self.player.scom.unofficial_set_play_mode("GameOver")` at the end of the gym’s `__init__` method.
> 3. **Setting the play mode to `PlayOn`** using `self.player.scom.unofficial_set_play_mode("PlayOn")` at the end of the gym’s `__init__` method, and then frequently resetting the game time (e.g., at every step) using `self.player.scom.unofficial_set_game_time(0)`. This option checks for self-collisions, which can be useful for certain behaviors (e.g. kicking the ball). However, note that the referee will still intervene in cases like the ball going out of bounds or entering a goal.
> 
> ### Code example to refer to
> You can use `/scripts/gyms/Basic_Run.py` as a base behavior for your implementation.
> 
> ### Dividing training into two steps
> Your approach of splitting the training into two steps (walking towards the ball and then kicking) is a good idea. However, it's crucial to organize your `reset()` and `step()` functions correctly:
> 
> 1. **`reset()` Function**:
>    
>    * In the `reset()` function, you should handle all the non-learned behaviors:
>      
>      * Place the ball at a specific position.
>      * Program the robot to walk towards the ball.
>      * End the `reset()` function when the robot is close enough to kick the ball.
>    * This ensures that the learning process focuses solely on the kicking part during the `step()` function.
> 2. **`step()` Function**:
>    
>    * Here is a draft implementation of the `step()` function, assuming an action size of 16 (defined in `__init__` as `self.no_of_actions = act_size = 16`):
>      def step(self, action):
>          w = self.player.world
>          r = self.player.world.robot
>      
>          # Example rescaling of the action outputs for a kick
>          r.joints_target_speed[2:18] = action * [6, 6, 3, 3, 6, 6, 6, 6, 6, 6, 2, 2, 1, 1, 1, 1]
>          
>          self.sync()  # Run simulation step
>          self.step_counter += 1
>          
>          if self.step_counter > 20:  # Allow 0.4 seconds for the kick (adjust as needed)
>              # Stop controlling the robot and evaluate the kick
>              while True:
>                  # (CODE TO EXIT LOOP WHEN THE BALL STOPS)
>                  r.behavior.execute("Zero_Bent_Knees")  # Example idle behavior
>                  self.sync()  # Run simulation step
>      
>              # (CODE TO CALCULATE YOUR_REWARD)            
>      
>              # Terminate the episode and return the reward
>              return self.observe(), YOUR_REWARD, True, {}
>          
>          else:  # No reward until the end
>              return self.observe(), 0, False, {}
> 
> ### Adjusting the number of steps per environment
> Since the episodes in your environment are very short, you might want to adjust the `n_steps_per_env` hyperparameter in the `train()` function of the `Train` class. For example, you can set:
> 
> n_steps_per_env = 128



I want to train the kicking behavior into the XML format in the slot. How should I do it?

**Comment by m-abr** [comment]
To give some context to my earlier answer, I explained how to train a kicking behavior from scratch. This method avoids relying too much on human-created data, which means less bias and more flexibility since it’s not limited by the slot structure in the XML. However, it can be tricky, especially if you want the kick to have a preparation phase, like swinging the foot back for a stronger kick. Sometimes, the policy learned from scratch finds a faster way to kick and doesn’t explore other types of kicks unless you add extra incentives.

If you’d rather improve an existing kick motion instead of starting from scratch, check out the GitHub issue **How to optimize your kicking motion** #5 . It provides an overview of how the `Get_Up.py` gym works and explains how to edit the XML files to adjust or add keyframes.

**Comment by zkk2157** [comment]
What aspects of the reward value should I consider in the optimization of kicking action?

**Comment by m-abr** [comment]
There are many ways to design rewards for optimizing a kick, and the best approach depends on your goals. Here are just a few ideas:

1. **Targeted Kicks**  
   If the goal is to hit a specific target, design a reward based on accuracy. A simple option is the inverse of the distance between the ball's stopping point and the target.

2. **Scoring Goals**  
   For learning to score, focus on rewarding the direction and strength of the kick. You can wait for the ball to stop and check its final position, but another method is to measure the ball’s speed a few time steps after the kick and reward based on speed and direction.

3. **Avoiding Falls**  
   For shorter kicks, you might add a reward for maintaining balance after kicking the ball. However, if falling isn’t an issue, it’s best to skip this to avoid adding unnecessary complexity to the training.

4. **Fast Kicks**  
   If speed is a priority, include a reward for how quickly the ball is kicked. However, avoid this if precision is your focus, as fast kicks can reduce accuracy. Alternatively, you can control kick speed by limiting the episode length.  

   With slot-based behaviors, you can independently limit the duration of each slot. For example, in **Kick_Motion.xml**, there are two slots: one for leaning forward and one for kicking. You can limit one or both slots, but note that this adds bias to the already constrained slot-based structure.  

   Here’s an example of how to limit the second slot to 0.5 seconds in `Get_Up.py`:

```python
if self.current_slot == 1:
    new_delta = min(max((delta + action[0]) // 20 * 20, 20), 500)
else:
    # This is the original line for other slots
    new_delta = max((delta + action[0]) // 20 * 20, 20)

**Comment by zkk2157** [comment]
Thanks a lot for your help

**Comment by zkk2157** [comment]
> ### 处理球重置到场地中间
> 这种行为很可能是由于裁判干扰了球的位置造成的。如文档（[链接在此处](https://docs.google.com/document/d/1aJhwK2iJtU-ri_2JOB8iYvxzbPskJ8kbk_4rb3IK3yc/edit?tab=t.0#heading=h.txmrqxm6q5b3)）中所述，在第 _10.4.1 节 Referee 干扰_下，您可以通过选择以下选项之一来处理此问题：
> 
> 1. 通过服务器的配置参数**关闭 referee** （但请注意，由于服务器 bug，这目前可能不起作用）。
> 2. **将播放模式设置为 ** using at the end of the gym 的方法。`self.player.scom.unofficial_set_play_mode("GameOver")``__init__`
> 3. **将游戏模式设置为**在健身房方法结束时使用 ，然后使用 频繁重置游戏时间（例如，在每一步）。此选项检查自碰撞，这对于某些行为（例如踢球）非常有用。但是，请注意，裁判仍会在球出界或进入球门等情况下进行干预。`self.player.scom.unofficial_set_play_mode("PlayOn")``__init__``self.player.scom.unofficial_set_game_time(0)`
> 
> ### 要参考的代码示例
> 您可以用作实现的基本行为。`/scripts/gyms/Basic_Run.py`
> 
> ### 将培训分为两个步骤
> 您将训练分为两个步骤（走向球然后踢球）的方法是一个好主意。但是，正确组织 and 函数至关重要：`reset()``step()`
> 
> 1. **功能**：
>    
>    * 在该函数中，您应该处理所有未学习的行为：`reset()`
>      * 将球放在特定位置。
>      * 对机器人进行编程，使其走向球。
>      * 当机器人足够接近以踢球时结束该功能。`reset()`
>    * 这确保了学习过程只集中在功能期间的踢球部分。`step()`
> 2. **功能**：
>    
>    * 以下是该函数的草案实现，假设作大小为 16（定义为 ）：`step()``__init__``self.no_of_actions = act_size = 16`
>      def step(self, action):
>          w = self.player.world
>          r = self.player.world.robot
>      
>          # Example rescaling of the action outputs for a kick
>          r.joints_target_speed[2:18] = action * [6, 6, 3, 3, 6, 6, 6, 6, 6, 6, 2, 2, 1, 1, 1, 1]
>          
>          self.sync()  # Run simulation step
>          self.step_counter += 1
>          
>          if self.step_counter > 20:  # Allow 0.4 seconds for the kick (adjust as needed)
>              # Stop controlling the robot and evaluate the kick
>              while True:
>                  # (CODE TO EXIT LOOP WHEN THE BALL STOPS)
>                  r.behavior.execute("Zero_Bent_Knees")  # Example idle behavior
>                  self.sync()  # Run simulation step
>      
>              # (CODE TO CALCULATE YOUR_REWARD)            
>      
>              # Terminate the episode and return the reward
>              return self.observe(), YOUR_REWARD, True, {}
>          
>          else:  # No reward until the end
>              return self.observe(), 0, False, {}
> 
> ### 调整每个环境的步骤数
> 由于环境中的剧集非常短，因此您可能需要调整类函数中的超参数。例如，您可以设置：`n_steps_per_env``train()``Train`
> 
> n_steps_per_env = 128

I used this method to train the kicking behavior. I added a small backswing behavior to the agent (larger backswing makes it hard to keep balance). After 3 million time steps of training, the kicking distance has been around 7.5m. I printed the action in step(), and the action value is very large. I tried to adjust the scale to the first 6 time steps of adding the backswing, but after 3 million time steps of training, the action value is still very large. Is this because my backswing is not perfect enough?

**Comment by m-abr** [comment]
An imperfect backswing wouldn’t be my first guess for the cause of this issue. One thing you could try is adding a very small penalty for action size in the reward function and seeing if the values still grow. I mentioned this approach in #39, it should help determine whether the policy is outputting large values simply due to a lack of incentive to keep them small. If the penalty doesn’t help, we can investigate further.

**Comment by zkk2157** [comment]
> An imperfect backswing wouldn’t be my first guess for the cause of this issue. One thing you could try is adding a very small penalty for action size in the reward function and seeing if the values still grow. I mentioned this approach in [#39](https://github.com/m-abr/FCPCodebase/issues/30#39), it should help determine whether the policy is outputting large values simply due to a lack of incentive to keep them small. If the penalty doesn’t help, we can investigate further.

Thank you for your help. I used the method you provided. After running 20 million time steps, I checked the value of action again. It was not very large and should be within the normal range. However, the kicking distance has been between 12 and 13 meters and has not increased.

**Comment by zkk2157** [comment]
I checked his kicking behavior and found that his backswing leg was not strong enough and the swing amplitude was not high enough. If he continued to train, could he learn a more perfect backswing?

**Comment by m-abr** [comment]
Yes, it’s a good idea to continue training for more than 20 million time steps to see how the behavior develops. Since the backswing is part of the overall kicking behavior, it can improve as the agent learns a better kick.

**Comment by zkk2157** [comment]
> 是的，最好继续训练超过 2000 万个时间步长，以了解行为如何发展。由于后摆是整体踢球行为的一部分，因此它可以随着代理学习更好的踢球而改善。

Your paper states that each time the ball is reset, the position of the ball is randomly placed, the agent is also randomly placed near the ball, and the direction of the kick is fixed. My approach is to keep the positions of the ball and the agent fixed. Will this have any impact?

**Comment by m-abr** [comment]
Randomizing the ball and agent positions helps the policy generalize better to real game scenarios. It forces the robot to approach the ball and balance itself from different angles and positions, which creates slight variations in each episode's initial state.

It also helps mitigate self-localization variations, as the agent produces localization data with varying levels of accuracy and noise depending on its position on the field—due to differences in visual landmark distance. This is especially important when kicking toward a specific target.

**Comment by zkk2157** [comment]
Thank you for your answer. I am currently training in this way, but the training speed is very slow, which may be due to insufficient CPU performance.

**Comment by qweasd982** [comment]
> 我不认为这个问题的首要原因是后摆不完美。你可以尝试在奖励函数中对动作大小添加一个非常小的惩罚，看看值是否还会增长。我在[#39](https://github.com/m-abr/FCPCodebase/issues/30#39)中提到过这种方法，它应该有助于确定策略输出较大的值是否仅仅是因为缺乏保持较小值的激励。如果惩罚没有帮助，我们可以进一步调查。

I also encountered the same problem. The value of action is very large. I want to know what the reason is. Is it caused by unreasonable scaling?

**Comment by m-abr** [comment]
I am not sure what is causing the large values. As I mentioned in #39:

> A simple thing to try is adding a very small penalty for action size in the reward function and checking if the values still grow. This should help because, in theory, the policy has no reason to output very large values unless the EMA filter (applied to the policy's action in Basic_Run.py) is too aggressive. If the penalty stops the actions from increasing, then the issue was just a lack of incentive to keep them small. If not, we can look deeper.

**Comment by qweasd982** [comment]
I found my problem. In order to meet the hyperparameter settings in the paper table, I set the following hyperparameters.

![Image](https://github.com/user-attachments/assets/301e3fff-72f0-490a-b3e8-d7b78cf8e8f9)

> 我不确定是什么导致了这些值这么大。正如我在[#39](https://github.com/m-abr/FCPCodebase/issues/39)中提到的：
> 
> > 一个简单的尝试是在奖励函数中对动作大小添加一个非常小的惩罚，并检查值是否仍在增长。这应该会有所帮助，因为理论上，除非EMA过滤器（应用于Basic_Run.py中策略的动作）过于激进，否则策略没有理由输出非常大的值。如果惩罚阻止了动作的增加，那么问题就在于缺乏激励来保持较小的值。如果不是，我们可以更深入地研究。
